Add config level for worktrees

vermiculus opened the pull request:
This uses the new support in libgit2 for [the worktree-config extension][1].

[1]: https://git-scm.com/docs/git-worktree/2.35.1#_configuration_file

See also https://github.com/libgit2/libgit2/pull/6202.

---

I suspect there is more work to be done on this PR, but I wanted to create it for tracking purposes. Tasks I can tell are coming:

- [ ] once the upstream PR is merged, update the libgit2 submodule
- [x] add unit tests around worktree-level config
  - no unit tests current exist for checking different levels of config
- [x] determine if documentation needs updating
  - no special documentation exists mentioning the `ConfigLevel` enum
- [x] determine if any libgit2/git2-rs glue code needs updating

vermiculus committed:
Add config level for worktrees

Use the new support in libgit2 for the worktree-config extension.[1]

[1]: https://git-scm.com/docs/git-worktree/2.35.1#_configuration_file

vermiculus commented:
Putting this back on the radar. libgit2/libgit2#6202 is passing tests and out for review; I'm hoping it will be merged in the near future.

vermiculus committed:
WIP Use my libgit2 patch pre-merge

vermiculus committed:
WIP fix word size issue? Is this macOS-specific?

ehuss commented:
For the ssh memory credential issue, the build.rs will need to be updated to accommodate https://github.com/libgit2/libgit2/pull/6617. In particular, it looks like GIT_SSH_MEMORY_CREDENTIALS was renamed to GIT_SSH_LIBSSH2_MEMORY_CREDENTIALS.

The test is also seg-faulting. Here's a backtrace:

```
* thread #36, name = 'remote::tests::push', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
    frame #0: 0x000000010024a1e4 git2-6f482251d0c12a64`git_remote_upload(remote=0x0000000121e047a0, refspecs=0x0000000171271ce0, opts=0x0000000171271de0) at remote.c:2987:31
   2984
   2985		if (opts && opts->remote_push_options.count > 0)
   2986			for (i = 0; i < opts->remote_push_options.count; ++i) {
-> 2987				char *optstr = git__strdup(opts->remote_push_options.strings[i]);
   2988				GIT_ERROR_CHECK_ALLOC(optstr);
   2989
   2990				if ((error = git_vector_insert(&push->remote_push_options, optstr)) < 0)
Target 0: (git2-6f482251d0c12a64) stopped.
(lldb) bt
* thread #36, name = 'remote::tests::push', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
  * frame #0: 0x000000010024a1e4 git2-6f482251d0c12a64`git_remote_upload(remote=0x0000000121e047a0, refspecs=0x0000000171271ce0, opts=0x0000000171271de0) at remote.c:2987:31
    frame #1: 0x000000010024a4b4 git2-6f482251d0c12a64`git_remote_push(remote=0x0000000121e047a0, refspecs=0x0000000171271ce0, opts=0x0000000171271de0) at remote.c:3025:15
    frame #2: 0x000000010009ca74 git2-6f482251d0c12a64`git2::remote::Remote::push::h68a78a039f76ea65(self=0x00000001712721d8, refspecs=&[&str] @ 0x0000000171271f20, opts=Option<&mut git2::remote::PushOptions> @ 0x0000000171271f30) at remote.rs:355:13
    frame #3: 0x00000001000468d8 git2-6f482251d0c12a64`git2::remote::tests::push::h32879140cc745f01 at remote.rs:979:13
    frame #4: 0x0000000100105884 git2-6f482251d0c12a64`git2::remote::tests::push::_$u7b$$u7b$closure$u7d$$u7d$::h87f24d99dd0e3fbb((null)=0x000000017127279e) at remote.rs:956:14
```

vermiculus committed:
Add new field for push-options

To fix a segfault at [1], adjust to upstream development[2,3] that
changed the shape of this struct.

This new field is added in the same style as `custom_headers`.

[1]: libgit2.git:39669956fb510fb7b13289f6ce959884969dbebd:src/libgit2/remote.c:2987
[2]: libgit2.git:ecc6f2fb8399d84e5b2bf043376dfc51f43f3e90
[3]: libgit2.git:39669956fb510fb7b13289f6ce959884969dbebd

vermiculus committed:
Use new name of GIT_SSH_MEMORY_CREDENTIALS

Adjust to upstream development[1] that changed the name of this
compiler flag.

[1]: libgit2.git:3fccf7468ba7a9a9f764984435db83f4588ec905

vermiculus commented:
@ehuss Thanks for the info; the segfault has now been fixed in 7c11a0f124d00da8a1d751cca65143e8b8b3982d.

If it's not too much trouble, how did you get that backtrace? `RUST_BACKTRACE=1 cargo test` offered no such insight.

ehuss commented:
> If it's not too much trouble, how did you get that backtrace? `RUST_BACKTRACE=1 cargo test` offered no such insight.

I used `lldb` on the test executable.
